src/Interpreters/DatabaseCatalog.cpp

```diff
@@ -930,6 +939,100 @@ void DatabaseCatalog::enqueueDroppedTableCleanup(StorageID table_id, StoragePtr
         (*drop_task)->schedule();
 }
 
+void DatabaseCatalog::dequeueDroppedTableCleanup(StorageID table_id)
+{
+    String latest_metadata_dropped_path;
+    String table_metadata_path;
+    StorageID dropped_table_id = table_id;
+    TablesMarkedAsDropped::iterator table;
+    {
+        std::lock_guard lock(tables_marked_dropped_mutex);
+        table = tables_marked_dropped.end();
+        time_t latest_drop_time = std::numeric_limits<time_t>::min();
+        for (auto it = tables_marked_dropped.begin(); it != tables_marked_dropped.end(); ++it)
+        {
+            if (it->table_id.uuid == table_id.uuid)
+            {
+                table = it;
+                break;
+            }
+
+            if (it->table_id.database_name == table_id.database_name &&
+                it->table_id.table_name == table_id.table_name &&
+                it->drop_time >= latest_drop_time)
+            {
+                latest_drop_time = it->drop_time;
+                table = it;
+            }
+        }
+
+        if (table == tables_marked_dropped.end())
+            throw Exception(ErrorCodes::UNKNOWN_TABLE,
+                "The drop task of table {} is in progress, has been dropped or the database engine doesn't support it",
+                table_id.getNameForLogs());
+
+        latest_metadata_dropped_path = table->metadata_path;
+        dropped_table_id = table->table_id;
+        table_metadata_path = getPathForMetadata(dropped_table_id);
+
+        fs::rename(latest_metadata_dropped_path, table_metadata_path);
+
+        tables_marked_dropped.erase(table);
+        if (tables_marked_dropped_ids.contains(dropped_table_id.uuid))
+            tables_marked_dropped_ids.erase(dropped_table_id.uuid);
```

Comment:
don't think we can have table in `tables_marked_dropped` and not in `tables_marked_dropped_ids` so this can be assert on `erase` return value.